packages/alchemy/src/middleware/gas-manager.ts

```diff
@@ -13,35 +13,65 @@ export interface AlchemyGasManagerConfig {
 /**
  * This middleware wraps the Alchemy Gas Manager APIs to provide more flexible UserOperation gas sponsorship.
  *
- * If `estimateGas` is true, it will use `alchemy_requestGasAndPaymasterAndData` to get all of the gas estimates + paymaster data
+ * If `delegateGasEstimation` is true, it will use `alchemy_requestGasAndPaymasterAndData` to get all of the gas estimates + paymaster data
  * in one RPC call.
  *
  * Otherwise, it will use `alchemy_requestPaymasterAndData` to get only paymaster data, allowing you
  * to customize the gas and fee estimation middleware.
  *
  * @param provider - the smart account provider to override to use the alchemy gas manager
  * @param config - the alchemy gas manager configuration
- * @param estimateGas - if true, this will use `alchemy_requestGasAndPaymasterAndData` else will use `alchemy_requestPaymasterAndData`
+ * @param delegateGasEstimation - if true, this will use `alchemy_requestGasAndPaymasterAndData` else will use `alchemy_requestPaymasterAndData`
  * @returns the provider augmented to use the alchemy gas manager
  */
 export const withAlchemyGasManager = <P extends AlchemyProvider>(
   provider: P,
   config: AlchemyGasManagerConfig,
-  estimateGas: boolean = true
+  delegateGasEstimation: boolean = true
 ): P => {
-  return estimateGas
+  const fallbackGasEstimator = provider.gasEstimator;
+  const fallbackFeeDataGetter = provider.feeDataGetter;
+
+  return delegateGasEstimation
     ? provider
         // no-op gas estimator
-        .withGasEstimator(async () => ({
-          callGasLimit: 0n,
-          preVerificationGas: 0n,
-          verificationGasLimit: 0n,
-        }))
+        .withGasEstimator(async (struct, overrides) => {
+          // but if user is bypassing paymaster to fallback to having the account to pay the gas (one-off override),
+          // we cannot delegate gas estimation to the bundler because paymaster middleware will not be called
+          if (overrides?.paymasterAndData !== undefined) {
```

Comment:
I think we shouldn't allow any `paymasterAndData` override here and only allow someone to set it to `0x` (as in skip paymaster for this UO) if the paymaster and data expects a specific UO hash, then reestimating gas will result in a broken UO.

Reply:
discussed offline